cmd/load-generator/state.go

```diff
@@ -0,0 +1,472 @@
+				Timeout:   10 * time.Second,
+				KeepAlive: 30 * time.Second,
+			}).DialContext,
+			TLSHandshakeTimeout: 5 * time.Second,
+			TLSClientConfig: &tls.Config{
+				InsecureSkipVerify: true, // CDN bypass can cause validation failures
+			},
+			MaxIdleConns:    500,
+			IdleConnTimeout: 90 * time.Second,
+		},
+		Timeout: 10 * time.Second,
+	}
+	latencyFile, err := newLatencyFile(latencyPath)
+	if err != nil {
+		return nil, err
+	}
+	s := &State{
+		client:      client,
+		apiBase:     apiBase,
+		certKey:     certKey,
+		domainBase:  domainBase,
+		callLatency: latencyFile,
+		wg:          new(sync.WaitGroup),
+		realIP:      realIP,
+		maxRegs:     maxRegs,
+		email:       userEmail,
+		respCodes:   make(map[int]*respCode),
+	}
+
+	// convert operations strings to methods
+	for _, opName := range operations {
+		op, present := stringToOperation[opName]
+		if !present {
+			return nil, fmt.Errorf("unknown operation %q", opName)
+		}
+		s.operations = append(s.operations, op)
+	}
+
+	return s, nil
+}
+
+// Run runs the WFE load-generator
+func (s *State) Run(httpOneAddr string, tlsOneAddr string, p Plan) error {
+	s.challSrv = newChallSrv(httpOneAddr, tlsOneAddr)
+	s.challSrv.run()
+	fmt.Printf("[+] Started challenge servers, http-01: %q, tls-sni-01: %q\n", httpOneAddr, tlsOneAddr)
+
+	if p.Delta != nil {
+		go func() {
+			for {
+				time.Sleep(p.Delta.Period)
+				atomic.AddInt64(&p.Rate, p.Delta.Inc)
+			}
+		}()
+	}
+
+	// Run sending loop
+	stop := make(chan bool, 1)
+	sigs := make(chan os.Signal, 1)
+	signal.Notify(sigs, syscall.SIGINT, syscall.SIGTERM)
+	fmt.Println("[+] Beginning execution plan")
+	i := int64(0)
+	go func() {
+		for {
+			start := time.Now()
+			select {
+			case <-stop:
+				return
+			default:
+				s.wg.Add(1)
+				go s.sendCall()
+				atomic.AddInt64(&i, 1)
+			}
+			sf := time.Duration(time.Second.Nanoseconds()/atomic.LoadInt64(&p.Rate)) - time.Since(start)
+			time.Sleep(sf)
+		}
+	}()
+	go func() {
+		lastTotal := int64(0)
+		lastGet := int64(0)
+		lastPost := int64(0)
+		for {
+			time.Sleep(time.Second)
+			curTotal := atomic.LoadInt64(&i)
+			curGet := atomic.LoadInt64(&s.getTotal)
+			curPost := atomic.LoadInt64(&s.postTotal)
+			fmt.Printf(
+				"%s Action rate: %d/s [expected: %d/s], Request rate: %d/s [POST: %d/s, GET: %d/s], Responses: [%s]\n",
+				time.Now().Format("2006-01-02 15:04:05"),
+				curTotal-lastTotal,
+				atomic.LoadInt64(&p.Rate),
+				(curGet+curPost)-(lastGet+lastPost),
+				curGet-lastGet,
+				curPost-lastPost,
+				s.respCodeString(),
+			)
+			lastTotal = curTotal
+			lastGet = curGet
+			lastPost = curPost
+		}
+	}()
+
+	select {
+	case <-time.After(p.Runtime):
+		fmt.Println("[+] Execution plan finished")
+	case sig := <-sigs:
+		fmt.Printf("[!] Execution plan interrupted: %s caught\n", sig.String())
+	}
+	stop <- true
+	fmt.Println("[+] Waiting for pending flows to finish before killing challenge server")
+	s.wg.Wait()
+	return nil
+}
+
+// HTTP utils
```

Comment:
these might be a good candidate for a separate `.go` in the same package. Thoughts?